### DISCUSSION TITLE
Delete Record from Self-Referencing Model

### DISCUSSION THREAD

**Discussion by tony-cocco** [prisma/prisma#14505] [Q&A]
I have a model of some hierarchical data: a person has children and parents both of type person. If I call `prisma.person.delete({ id: person1 })` I get a foreign key constraint failure since any parents have a reference to person1, as well as any children. Is there a way to call delete so I don't have to disconnect all the records first?

An example schema looks like:

```
model Person {
  id        Int     @id @default(autoincrement())
  name      String?
  parents   Person[]  @relation("FamilyTree")
  children  Person[]  @relation("FamilyTree")
}
```

The following does work, but seems excessively verbose:
```
// is this the best way?
await db.person.update({
  where: { id: 'person1' },
  data: {
  parents: {
    set: [],
  },
  children: {
    set: [],
  },
});
await db.person.delete({where: { id: 'person1' });
```

For reference, this is the documentation I'm basing my model off of: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/self-relations#many-to-many-self-relations using SQL Server as my db provide.

**Comment by nurul3101** [comment]
Hey @tony-cocco 👋 

You mentioned that your data is hierarchical in nature, so Referential Actions might be useful for you to control the behaviour of how a referenced record is handled by the database when you run an [update](https://www.prisma.io/docs/concepts/components/prisma-client/crud#update) or [delete](https://www.prisma.io/docs/concepts/components/prisma-client/crud#delete) query.

Here are the default Referential Actions which Prisma uses: [Default Referential Actions](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#referential-action-defaults)

In your case you are using Implicit Many to Many relations in the `Person` Model, Referential actions are not supported in Implicit many to many relations, so if you would use explicit many to many to define relations then you should be able to control the delete behaviour for your referenced record.

Also, the approach that you are using is documented [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#disconnect-all-related-records) and you could certainly disconnect related records before deleting the actual record.

**Reply by tony-cocco** [reply]
Hey @nurul3101 thanks for your answer. I did try going down the route of the explicit many-to-many and using referential actions. I think the disconnect approach is better for our current codebase. I was hoping there was some nice syntax-sugar I didn't know about.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
